File: Practica1/Ejercicio2/Ejercicio2/lib/Carrera.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


class Carrera
  
  def initialize(porcentaje)
    
      @bicicletas = Array.new
      @porcentajeAbandono = porcentaje
    
  end
  
  #def self.new()
  #  raise "Error:Clase Abstracta"
  #end
  
  def aniadirbicicleta (bicicleta)
    
    id = @bicicletas.length
    #bicicleta.setidentificador(id)
    bicicleta.identificador = id
    
    #@bicicletas.push(bicicleta)
    @bicicletas << bicicleta
    
  end
  
  def abandonarcarrera
    
    abandonan = ( ( @porcentajeAbandono * (@bicicletas.length) ) / 100 )
    
    abandonan.times do |num|
      siguiente = rand(@bicicletas.count)
      @bicicletas.delete_at(siguiente)
    end
    
    return abandonan
    
  end
  
  def clone
    raise "Abstracto"
  end
  
  def run
    mensaje = "Iniciando carrera de " + printtipo + " con bicicletas:"
    
    #@bicicletas.size.times do |num|
    #  mensaje += " #{@bicicletas[num].getidentificador}"
    #end
    
    #for bicicleta in @bicicletas
    #  mensaje += " #{bicicleta.identificador}"
    #end
    
    @bicicletas.each { |bicicleta|
      mensaje += " #{bicicleta.identificador}"
    }
    
    puts mensaje
    
    sleep(5)
    
    numAbandonos = abandonarcarrera()
    
    puts "Carrera de " + printtipo + " - han abandonando #{numAbandonos} bicicletas"
    
    sleep(5)
    
    mensaje = "Carrera de " + printtipo + " terminada con biciletas: "
    
    #@bicicletas.size.times do |num|
    #  mensaje += " #{@bicicletas[num].getidentificador}"
    #end
    
    @bicicletas.each { |bicicleta|
      mensaje += " #{bicicleta.identificador}"
    }
    
    puts mensaje
    
  end
  
end

File: Practica1/Ejercicio1/src/Carrera.java

import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author juanfrandm98
 */
public abstract class Carrera extends Thread {
    
    private ArrayList<Bicicleta> bicicletas;
    private int porcentajeAbandono;
    
    public Carrera( int porcentajeAbandono ) {
        bicicletas = new ArrayList();
        this.porcentajeAbandono = porcentajeAbandono;
    }
    
    public void aniadirBicicleta( Bicicleta bicicleta ) {
        
        int nuevoIdentificador = bicicletas.size();
        bicicleta.setIdentificador( nuevoIdentificador );
        
        bicicletas.add( bicicleta );
        
    }
    public int abandonarCarrera(){
        
        int abandonan = (this.porcentajeAbandono * bicicletas.size())/100;
        
        for(int i = 0 ; i < abandonan ; i++){
            
            int siguiente = (int) (Math.random() * bicicletas.size() );
            bicicletas.remove( siguiente );
            
        }
        
        return abandonan;
    }
    
    @Override
    public void run() {
        
        String mensaje = "Iniciando carrera de " + printTipo() + " con bicicletas:";
        
        for( Bicicleta b : bicicletas )
            mensaje += " " + b.getIdentificador();
        
        System.out.println( mensaje );
        
        try {
            sleep(5000);
        } catch (InterruptedException ex) {
            Logger.getLogger(Carrera.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        int numAbandonos = abandonarCarrera();
        
        System.out.println( "Carrera " + printTipo() + " - han abandonado " + numAbandonos + " bicicletas." );
        
        try {
            sleep(5000);
        } catch (InterruptedException ex) {
            Logger.getLogger(Carrera.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        mensaje = "Carrera de " + printTipo() + " terminada con bicicletas:";
        
        for( Bicicleta b : bicicletas )
            mensaje += " " + b.getIdentificador();
        
        System.out.println( mensaje );
        
    }
    
    public abstract String printTipo();
    
}
